File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonGenericMessageDeserializer.java
package com.github.dedis.popstellar.model.network.serializer;

import com.github.dedis.popstellar.model.network.GenericMessage;
import com.github.dedis.popstellar.model.network.answer.Answer;
import com.github.dedis.popstellar.model.network.method.Message;
import com.google.gson.*;

import java.lang.reflect.Type;

import timber.log.Timber;

/** Json deserializer for the generic messages */
public class JsonGenericMessageDeserializer implements JsonDeserializer<GenericMessage> {
  public static final String TAG = JsonGenericMessageDeserializer.class.getSimpleName();
  private static final String METHOD = "method";

  @Override
  public GenericMessage deserialize(
      JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    Timber.tag(TAG).d("deserializing generic message");
    if (json.getAsJsonObject().has(METHOD)) {
      return context.deserialize(json, Message.class);
    } else {
      return context.deserialize(json, Answer.class);
    }
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/network/JsonGenericMessageDeserializer.kt
package com.github.dedis.popstellar.model.network.serializer.network

import com.github.dedis.popstellar.model.network.GenericMessage
import com.github.dedis.popstellar.model.network.answer.Answer
import com.github.dedis.popstellar.model.network.method.Message
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type
import timber.log.Timber

/** Json deserializer for the generic messages */
class JsonGenericMessageDeserializer : JsonDeserializer<GenericMessage> {
  @Throws(JsonParseException::class)
  override fun deserialize(
      json: JsonElement,
      typeOfT: Type,
      context: JsonDeserializationContext
  ): GenericMessage {
    Timber.tag(TAG).d("deserializing generic message")

    return if (json.asJsonObject.has(METHOD)) {
      context.deserialize(json, Message::class.java)
    } else {
      context.deserialize(json, Answer::class.java)
    }
  }

  companion object {
    val TAG: String = JsonGenericMessageDeserializer::class.java.simpleName
    private const val METHOD = "method"
  }
}
